third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,242 @@
+        e: &ExpData,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        use ExpData::*;
+        use VisitorPosition::*;
+        match (position, e) {
+            (Pre, Call(nid, op, _)) => self.visit_call(nid, op, insts, callers_chain),
+            _ => true,
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`
+    fn visit_call(
+        &self,
+        nid: &NodeId,
+        op: &Operation,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        if let Operation::MoveFunction(mod_id, fun_id) = op {
+            let callee_uninst = mod_id.qualified_inst(*fun_id, self.get_inst(*nid));
+            let callee = callee_uninst.instantiate(&insts);
+            if *mod_id != self.mod_env.get_id() || self.def_not_recursive(callee.to_qualified_id())
+            {
+                // skips if callee from another module (since there is no cyclic module dependency),
+                // or if the callee is not recursive
+                true
+            } else {
+                for (_, ancester_caller) in callers_chain.iter() {
+                    if ancester_caller.to_qualified_id() == callee.to_qualified_id() {
+                        // we are checking for the root caller
+                        let (_, checking_for) = &callers_chain[0];
+                        if checking_for.to_qualified_id() != callee.to_qualified_id() {
+                            // check and report diagnostics when `callee` is checked
+                            // this happens when root caller `f` calls `g` which then calls `g` itself
+                            return true;
+                        } else {
+                            #[allow(clippy::collapsible_else_if)]
+                            if let Some(_ty_param) = callee
+                                .inst
+                                .iter()
+                                .filter_map(ty_properly_contains_ty_parameter)
+                                .next()
+                            {
+                                self.report_error(*nid, callee, callers_chain);
+                                return false;
+                            } else {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                self.visit_callees(*nid, callee, insts, callers_chain)
+            }
+        } else {
+            true
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`.
+    /// Precondition: `caller` defined in `self.mod_env`
+    fn visit_callees(
+        &self,
+        caller_node: NodeId,
+        caller: QualifiedInstId<FunId>,
+        insts: Vec<Type>,
```

Comment:
You only use `&insts` below, so just take a parameter of `&Vec<Type>` instead of a value.

Reply:
done